### PR TITLE
Registry quota reworked

### PR DESCRIPTION
**Pull request by miminar**
Covers:

- Removes size counting from quota evaluation
- `openshift.io/projectimagessize` removed completely
- `openshift.io/imagestreamsize` removed completely.
- `openshift.io/imagesize` removed from quota evaluation.
- Added a new resource `openshift.io/images` that allows to limit number of images tagged into a project.
- ~~Improves error messages of failed builds.~~

~~So instead of:~~

    Error: build error: Failed to push image: unknown: requested access to the resource is denied

~~It will now show:~~

    Error: build error: Failed to push image: Refusing to write blob exceeding quota: openshift.io/images limited to {4.000 DecimalSI} by {1.000 DecimalSI}~~

Improved error messages will come in another follow-up.

Does **NOT** cover:

- `openshift.io/imagesize` being checked in limit range admission plugin
- `openshift.io/imagesize` being checked in the registry

TODO:

- [x] - fix extended test
- [x] - remove `openshift.io/projectimagessize` as well

### CODE REVIEW

**Comment by miminar** [comment]
[test]

**Comment by miminar** [comment]
Found out an issue that needs to be addressed upstream. I'll submit there and update the commit message.

**Comment by miminar** [comment]
[test][extended:core(image resource quota)]

File: test/extended/images/quota_admission.go [hunk 1]
```diff
@@ -66,58 +87,235 @@ var _ = g.Describe("[images] openshift image resource quota", func() {
...
+		o.Expect(err).NotTo(o.HaveOccurred())
+		err = waitForLimitSync(oc, quotaName, rq.Spec.Hard)
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		/** Test referencing using DockerImage.
+		 *
+		 * TODO: this is now failing during the image import with following status:
+		 *   Reason:"Unauthorized", Message:"you may not have access to the Docker image \"172.30.163.69:5000/extended-test-imagequota-admission-6znr2-shared/src@sha256:bb08da7a0b99af72aebe0ead77009d28771a30dbe4d99f49bf3d29407c72e48f\""
+		 * Uncomment once resolved (issue #7985).
+		 * The import seems to be working when the source and destination namespaces are the same.
```
Comment (miminar) [inline test/extended/images/quota_admission.go]: Issue #7985.

File: pkg/quota/image/imagestreamtag_evaluator.go [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+	}
+}
+
+// imageStreamUsageComputer is a context object for use in SharedContextEvaluator.
+type imageStreamTagUsageComputer struct {
+	GenericImageStreamUsageComputer
+}
+
+// Usage returns a usage for an image stream tag.
+func (c *imageStreamTagUsageComputer) Usage(object runtime.Object) kapi.ResourceList {
```
Comment (miminar) [inline pkg/quota/image/imagestreamtag_evaluator.go]: @derekwaynecarr This may be interesting to you.

File: pkg/quota/image/imagestream_evaluator.go [hunk 3]
```diff
@@ -62,33 +87,60 @@ func imageStreamConstraintsFunc(required []kapi.ResourceName, object runtime.Obj
...
+	}
+}
+
+// imageStreamUsageComputerForAdmission is a context object for use in SharedContextEvaluator
+type imageStreamUsageComputerForAdmission struct {
+	GenericImageStreamUsageComputer
+}
+
+// Usage returns a usage for an image stream.
+func (c *imageStreamUsageComputerForAdmission) Usage(object runtime.Object) kapi.ResourceList {
```
Comment (miminar) [inline pkg/quota/image/imagestream_evaluator.go]: @derekwaynecarr And this.

**Comment by miminar** [comment]
Got successful run.  @pweil-, @kargakis could you please take a look?

File: pkg/quota/image/imagestream_evaluator_test.go [hunk 4]
```diff
@@ -429,16 +466,412 @@ func TestImageStreamEvaluatorUsageStats(t *testing.T) {
...
+				imageapi.ResourceProjectImagesSize: resource.MustParse("682"),
+				// avoid hitting quota limit when there's no change to the tags
+				imageapi.ResourceImageStreamSize: resource.MustParse("0"),
+				// there's no new image to admit
+				imageapi.ResourceImageSize: resource.MustParse("0"),
+			},
+		},
+
+		{
+			name: "adding two new tags",
```
Comment (kargakis) [inline pkg/quota/image/imagestream_evaluator_test.go]: you also delete latest here, right?

Comment (miminar) [inline pkg/quota/image/imagestream_evaluator_test.go]: Right, I'll correct the description.

File: pkg/quota/admission/resourcequota/admission.go [hunk 5]
```diff
@@ -54,7 +54,7 @@ func (a *originQuotaAdmission) Admit(as admission.Attributes) error {
 }
 
 func (a *originQuotaAdmission) SetOpenshiftClient(osClient osclient.Interface) {
-	registry := quota.NewRegistry(osClient)
+	registry := quota.NewRegistry(osClient, true)
```
Comment (pweil-) [inline pkg/quota/admission/resourcequota/admission.go]: @derekwaynecarr - I get the feeling we need some sort of lifecycle method that can be called now that we have setters.  If the controller needs to set up reflectors or anything else that relies on the OpenShift client it's probably going to happen here for now and kind of makes the method jumbled.

File: pkg/quota/image/imagestream_evaluator.go [hunk 6]
```diff
@@ -38,15 +42,52 @@ func NewImageStreamEvaluator(osClient osclient.Interface) kquota.Evaluator {
...
+		imageStreamConstraintsFunc,
+		makeImageStreamUsageComputerFactory(osClient),
+		imageapi.ResourceImageStreamSize,
+		imageapi.ResourceImageSize)
+}
+
+// NewImageStreamAdmissionEvaluator computes resource usage of ImageStreams in the context of admission
+// plugin.
+func NewImageStreamAdmissionEvaluator(osClient osclient.Interface) kquota.Evaluator {
+	computeResources := []kapi.ResourceName{
```
Comment (pweil-) [inline pkg/quota/image/imagestream_evaluator.go]: This method could just call `NewImageStreamEvaluator` and set only the fields it needs to change to reduce some of the duplicate code

File: pkg/quota/image/imagestream_evaluator.go [hunk 7]
```diff
@@ -84,11 +121,59 @@ func (c *imageStreamUsageComputer) Usage(object runtime.Object) kapi.ResourceLis
...
+// Usage returns a usage for an image stream.
+func (c *imageStreamUsageComputerForAdmission) Usage(object runtime.Object) kapi.ResourceList {
+	newIS, ok := object.(*imageapi.ImageStream)
+	if !ok {
+		return kapi.ResourceList{}
+	}
+
+	oldIS, err := c.getImageStream(newIS.Namespace, newIS.Name)
+	if err != nil {
+		if !kerrors.IsNotFound(err) && newIS.Namespace != "" {
```
Comment (pweil-) [inline pkg/quota/image/imagestream_evaluator.go]: are you intentionally swallowing the error if the ns is set?  If so, a comment would be good.

File: pkg/quota/image/imagestream_evaluator.go (hunk 6 above)
Comment (miminar) [inline pkg/quota/image/imagestream_evaluator.go]: Good idea, I'll do so.

File: pkg/quota/image/imagestream_evaluator.go (hunk 7 above)
Comment (miminar) [inline pkg/quota/image/imagestream_evaluator.go]: This block of code doesn't appear in the final commit. I'll squash this with the following one.

File: pkg/build/builder/cmd/builder.go [hunk 8]
```diff
@@ -26,7 +26,7 @@ import (
 )
 
 type builder interface {
-	Build(dockerClient bld.DockerClient, sock string, buildsClient client.BuildInterface, build *api.Build, gitClient bld.GitClient, cgLimits *s2iapi.CGroupLimits) error
+	Build(dockerClient bld.DockerClient, sock string, buildsClient client.BuildInterface, build *api.Build, gitClient bld.GitClient, eventsClient kclient.EventInterface, cgLimits *s2iapi.CGroupLimits) error
```
Comment (pweil-) [inline pkg/build/builder/cmd/builder.go]: I'd make this a new PR, it isn't directly related to the image quota rework

Comment (miminar) [inline pkg/build/builder/cmd/builder.go]: Alright.

File: pkg/quota/image/imagestream_evaluator.go (hunk 7 above)
Comment (pweil-) [inline pkg/quota/image/imagestream_evaluator.go]: Ok.  I think the easiest way to review this is if the commits are ordered so that we have

1.  the revert
2.  the upstreams
3.  the new reworked code which will now look like a new implementation

@kargakis - wdyt?

Comment (kargakis) [inline pkg/quota/image/imagestream_evaluator.go]: Agreed. Unfortunately github doesn't help a lot with the ordering of the commits

**Comment by miminar** [comment]
Addressed comments. Removed *improved error messages* from this PR.

**Comment by miminar** [comment]
We're green. Any other comments?

**Comment by pweil-** [comment]
None from me

@soltysh - please give this some :eyes: 
@smarterclayton - FYI

**Comment by pweil-** [comment]
Also, hold for merge post-rebase so the drop commit falls out.

File: pkg/quota/admission/resourcequota/admission.go (hunk 5 above)
Comment (derekwaynecarr) [inline pkg/quota/admission/resourcequota/admission.go]: @pweil- I am not sure I follow.  The `quota.NewRegistry` code is completely unique to OpenShift.  The Kubernetes expectation is that you just return a `kquota.Registry` that knows how to evaluate use for OpenShift specific API types.  Maybe this comment will become clearer to me when I look at the rest of this PR.

File: pkg/quota/registry.go [hunk 9]
```diff
@@ -8,6 +8,11 @@ import (
 )
 
 // NewRegistry returns a registry object that knows how to evaluate quota usage of OpenShift resources.
-func NewRegistry(osClient osclient.Interface) kquota.Registry {
-	return image.NewImageRegistry(osClient)
+// If the registry will be used in admission plugin, pass true to forAdmission.
+func NewRegistry(osClient osclient.Interface, forAdmission bool) kquota.Registry {
```
Comment (derekwaynecarr) [inline pkg/quota/registry.go]: This is not obvious to me yet, but I am trying to work @pweil- comment backwards.

Upstream in Kubernetes, we did not encounter the need to have a different `Evaluator` for use in admission control than for use in the background quota controller.  The goal was to keep them unified since we should ideally be counting usage the same in both places.

I need to look elsewhere why this was needed, but it would be nice to at least comment in code why you required this split.

File: pkg/quota/admission/resourcequota/admission.go (hunk 5 above)
Comment (pweil-) [inline pkg/quota/admission/resourcequota/admission.go]: @derekwaynecarr - I am the champ at confusing comments.  This isn't related to the `NewRegistry` call. 

The scenario is if I have an admission controller that uses an OpenShift client I cannot fully initialize the controller until it has the client.  So, for example if I need reflectors that are watching OpenShift resources I found myself putting that creation logic in the `SetOpenshiftClient` method.

It would just be nice if there was a life cycle method that could be called that is after any Set* method rather than embedding that logic here.  Holdover from java days I guess, seeing logic unrelated to a field mentioned in the Set* signature smells to me but there is no real harm done.

File: pkg/quota/image/registry.go [hunk 10]
```diff
@@ -12,11 +12,24 @@ import (
...
+	return &generic.GenericRegistry{
+		InternalEvaluators: map[unversioned.GroupKind]quota.Evaluator{
+			imageStream.GroupKind(): imageStream,
+		},
+	}
+}
+
+// NewImageRegistry returns a registry for quota evaluation of OpenShift resources related to images and image
+// streams.
+func NewImageRegistryForAdmission(osClient osclient.Interface) quota.Registry {
```
Comment (derekwaynecarr) [inline pkg/quota/image/registry.go]: Apologize as I leave a trail of bread crumb comments on this PR, but this is my next step in understanding the need to split on admission control.  It's not yet clear from here why this split is needed.  A comment here would really help explain why.  Right now, its not-obvious to me why you don't collapse the admission registry into the normal `NewImageRegistry` call.  I need to continue my investigation.

Comment (derekwaynecarr) [inline pkg/quota/image/registry.go]: So the best that I can gather is that this is because of the fan-out problem where we are trying to quota the number of references to a non-namespaced resource.  This is something we never hit upstream because the closest parallel would be quota storage associated with a PV, but a PVC has a 1:1 relationship with a PV, whereas we have a 1:N relationship here...

I still think it would really help to have comments all over the place explaining this split.

File: pkg/quota/admission/resourcequota/admission.go (hunk 5 above)
Comment (miminar) [inline pkg/quota/admission/resourcequota/admission.go]: :+1:

File: pkg/quota/image/imagestream_evaluator.go (hunk 7 above)
Comment (miminar) [inline pkg/quota/image/imagestream_evaluator.go]: Done.

File: pkg/quota/image/imagestreammapping_evaluator.go [hunk 11]
```diff
@@ -27,7 +27,7 @@ func NewImageStreamMappingEvaluator(osClient osclient.Interface) kquota.Evaluato
 	matchesScopeFunc := func(kapi.ResourceQuotaScope, runtime.Object) bool { return true }
 
 	return quotautil.NewSharedContextEvaluator(
-		"ImageStreamMapping admission evaluator",
+		"Evaluator.ImageStreamMapping.Admission",
```
Comment (soltysh) [inline pkg/quota/image/imagestreammapping_evaluator.go]: Can we have this and `Evaluator.ImageStream.ResourceQuotaController` being constants?

Comment (miminar) [inline pkg/quota/image/imagestreammapping_evaluator.go]: Sure, we can.

File: test/extended/images/quota_admission.go [hunk 12]
```diff
@@ -47,18 +38,48 @@ var _ = g.Describe("[images] openshift image resource quota", func() {
...
+			}
+			err := oc.AdminREST().ImageStreams(is.Namespace).Delete(is.Name)
+			o.Expect(err).NotTo(o.HaveOccurred())
+		}
+
+		g.By("Deleting shared project")
+		oc.AdminREST().Projects().Delete(oc.Namespace() + "-shared")
+	})
+
+	g.It("should deny a push of built image exceeding quota", func() {
```
Comment (soltysh) [inline test/extended/images/quota_admission.go]: should deny a push of ~~built~~ image exceeding quota 
It's more clear, the built is confusing here.

**Comment by soltysh** [comment]
Remove [UPSTREAM: 22766: <drop>: Fix validation of resources (cpu, memory, st… ](https://github.com/openshift/origin/pull/7930/commits/d14ef40219dbd6ee683df4ad0e4ba17ee27d4783) as @pweil- mentioned

[UPSTREAM: 22852: Set a missing namespace on objects to admit](https://github.com/openshift/origin/pull/7930/commits/0e74390c410dafe7d11e4e9a2c09e5d0375665c8) is not approved upstream

Plus some small nits. But generally I'd wait with merging this until upstream PR is good.

File: test/extended/images/quota_admission.go (hunk 12 above)
Comment (miminar) [inline test/extended/images/quota_admission.go]: Fixed.

File: pkg/quota/image/registry.go (hunk 10 above)
Comment (miminar) [inline pkg/quota/image/registry.go]: Comments added. Do you find them sufficient?

**Comment by miminar** [comment]
Comments addressed. Removed the upstream PR #22766.

File: pkg/quota/image/imagestreamtag_evaluator.go [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+	ist, ok := object.(*imageapi.ImageStreamTag)
+	if !ok {
+		return kapi.ResourceList{}
+	}
+
+	img := &ist.Image
```
Comment (miminar) [inline pkg/quota/image/imagestreamtag_evaluator.go]: Just found out this is a nonsense. We need to parse the `ist.Tag` object reference to get the right image. The `ist.Image` won't probably ever be set by the client. And even if it were, the master API won't pay attention to that.

Comment (smarterclayton) [inline pkg/quota/image/imagestreamtag_evaluator.go]: You're not going to know whether the image that is being tagged already exists in the image stream, or whether this is a duplicate tag.  This is partially why this check really can only be done in an admission controller inside of ImageStreamTag's registry (as it calls down to update the ImageStream).  I don't think this is going to lead to success, doing it in front of the ImageStreamTag.

Comment (miminar) [inline pkg/quota/image/imagestreamtag_evaluator.go]: I don't find it any different from handling an update of `ImageStream` having an `ImageStreamTag` reference in its spec. Again it's the weakest point of registry quota admission which you can work around by first tagging a non-existent or an image that already exist in the project and make the source *is* refer another new image afterwards.

For now the best-effort policy is to reject new image exceeding the quota if any and refuse anything else.

I wonder if we should attempt to address this by checking the quota of all dependent image streams (those referencing the *is* being updated) as well.

Anyway, I've updated the method to fetch and check an image referenced by `ist.Tag.From`.

**Comment by soltysh** [comment]
LGTM wait till upstream PR gets approval and merge.

**Comment by miminar** [comment]
Is it the extended test argument that suddently stopped working?

    real	17m55.263s
    user	0m5.266s
    sys	0m0.516s
    + vagrant test-origin --target 'test -o check' --extended 'core(image resource quota)' -d --skip-image-cleanup
    /var/lib/jenkins/.vagrant.d/gems/gems/vagrant-openshift-3.0.5/lib/vagrant-openshift/action/download_artifacts_origin.rb:24:in `initialize': wrong number of arguments (2 for 3) (ArgumentError)
            from /opt/vagrant/embedded/gems/gems/vagrant-1.7.4/lib/vagrant/action/warden.rb:90:in `new'
            from /opt/vagrant/embedded/gems/gems/vagrant-1.7.4/lib/vagrant/action/warden.rb:90:in `finalize_action'
            from /opt/vagrant/embedded/gems/gems/vagrant-1.7.4/lib/vagrant/action/warden.rb:20:in `block in initialize'
            from /opt/vagrant/embedded/gems/gems/vagrant-1.7.4/lib/vagrant/action/warden.rb:20:in `map'
            from /opt/vagrant/embedded/gems/gems/vagrant-1.7.4/lib/vagrant/action/warden.rb:20:in `initialize'
            from /opt/vagrant/embedded/gems/gems/vagrant-1.7.4/lib/vagrant/action/builder.rb:170:in `new'
            from /opt/vagrant/embedded/gems/gems/vagrant-1.7.4/lib/vagrant/action/builder.rb:170:in `to_app'
            from /opt/vagrant/embedded/gems/gems/vagrant-1.7.4/lib/vagrant/action/builder.rb:116:in `call'
    ...

**Comment by soltysh** [comment]
@miminar it splits the core tests on space, see [here](https://github.com/openshift/vagrant-openshift/blob/master/lib/vagrant-openshift/action/run_origin_tests.rb#L145) that's why you should avoid using spaces in the extended selector.

**Comment by miminar** [comment]
@soltysh [all the jobs](https://ci.openshift.redhat.com/jenkins/job/test_pull_requests_origin/) are failing with this error now. Not just those passing extended test args. I'm not sure who to ping ... it doesn't seem related to commit recently merged to origin.

**Comment by miminar** [comment]
Seems like it's been fixed. re-[test]

File: pkg/quota/image/imagestreamtag_evaluator.go (hunk 13 above)
Comment (smarterclayton) [inline pkg/quota/image/imagestreamtag_evaluator.go]: My point is that image streams are not generic blobs, but inputs that
are calculated after admission.  We should check the quota of the
image stream beforehand and then reject if you are over (which it
sounds like you agree with) but we can do that without attempting to
calculate the incoming tag change.

Tagging an image from another namespace or importing it is not a
serious problem for us if the image exceeds the quota (in both cases
size is accounted for).  Only the number of images per stream needs to
be enforced, and only the size of pushed images.  We don't even need
to be exact on the number of images per stream - if we are at 21 and
quota is 20, we're still fine.  Any create can be rejected, and any
delete can be allowed.

That is simpler than trying to do this incremental calculation, and
remove the need to duplicate business logic that only the image stream
should be aware of.

Basically, on create IST, check count of images and reject if over
limit.  Always allow IST delete.  Have replenish only look at IS.

**Comment by soltysh** [comment]
Upstream is merged, tests are green.
[merge]

File: pkg/quota/image/usage.go [hunk 14]
```diff
@@ -10,139 +13,400 @@ import (
...
 
-			if _, exists := processedImages[imgName]; exists {
-				continue
-			}
-			processedImages[imgName] = sets.Empty{}
+// NewGenericImageStreamUsageComputer returns an instance of GenericImageStreamUsageComputer.
+// Returned object can be used just once and must be thrown away afterwards.
+func NewGenericImageStreamUsageComputer(osClient osclient.Interface, processSpec bool, fetchImagesFromImageStream bool) *GenericImageStreamUsageComputer {
+	return &GenericImageStreamUsageComputer{
+		osClient,
```
Comment (smarterclayton) [inline pkg/quota/image/usage.go]: Always initialize structs with field names

File: pkg/quota/image/usage.go [hunk 15]
```diff
@@ -10,139 +13,400 @@ import (
...
+
+// getImageReferenceForObjectReference returns corresponding docker image reference for the given object
+// reference representing either an image stream image or image stream tag or docker image.
+func (c *GenericImageStreamUsageComputer) getImageReferenceForObjectReference(
+	namespace string,
+	objRef *kapi.ObjectReference,
+) (imageapi.DockerImageReference, error) {
+	switch objRef.Kind {
+	case "ImageStreamImage":
+		nameParts := strings.Split(objRef.Name, "@")
```
Comment (smarterclayton) [inline pkg/quota/image/usage.go]: Why are you parsing this twice?

File: pkg/quota/image/usage.go [hunk 16]
```diff
@@ -10,139 +13,400 @@ import (
...
+		if res.Namespace == "" {
+			res.Namespace = namespace
+		}
+		return res, nil
+
+	case "ImageStreamTag":
+		// This is really fishy. An admission check can be easily worked around by setting a tag reference
+		// to an ImageStreamTag with no or small image and then tagging a large image to the source tag.
+		// TODO: Shall we refuse an ImageStreamTag set in the spec if the quota is set?
+		nameParts := strings.Split(objRef.Name, ":")
```
Comment (smarterclayton) [inline pkg/quota/image/usage.go]: We already have a helper for this, you shouldn't be duplicating it.

File: pkg/quota/image/usage.go [hunk 17]
```diff
@@ -10,139 +13,400 @@ import (
...
+	}
+	return image, err
+}
+
+// listImageStreams returns a list of image streams of the given namespace and caches them for later access.
+func (c *GenericImageStreamUsageComputer) listImageStreams(namespace string) (*imageapi.ImageStreamList, error) {
+	iss, err := c.osClient.ImageStreams(namespace).List(kapi.ListOptions{})
+	if err == nil {
+		for _, is := range iss.Items {
+			c.imageStreamCache[fmt.Sprintf("%s/%s", namespace, is.Name)] = &is
```
Comment (smarterclayton) [inline pkg/quota/image/usage.go]: We already have cache constructs for this - why are we inventing another one?

File: pkg/quota/image/usage.go [hunk 18]
```diff
@@ -10,139 +13,400 @@ import (
...
+	}
+	return iss, err
+}
+
+// cacheInternalRegistryName caches registry name of the given docker image reference of an image stored in an
+// internal registry.
+func cacheInternalRegistryName(dockerImageReference string) {
+	ref, err := imageapi.ParseDockerImageReference(dockerImageReference)
+	if err == nil && len(ref.Registry) > 0 {
+		internalRegistryNames.Insert(ref.Registry)
```
Comment (smarterclayton) [inline pkg/quota/image/usage.go]: Don't use globals.

Comment (smarterclayton) [inline pkg/quota/image/usage.go]: This can lead to potentially unbounded memory use if we ever change the assumption that the internal registry is like this.  We should never be caching anything in globals anywhere in Origin code.

File: pkg/quota/image/imagestreamtag_evaluator.go [hunk 19]
```diff
@@ -0,0 +1,135 @@
...
+		getFuncByNamespace,
+		nil,
+		imageStreamTagConstraintsFunc,
+		makeImageStreamTagUsageComputerFactory(osClient))
+}
+
+// imageStreamTagConstraintsFunc checks that given object is an image stream tag
+func imageStreamTagConstraintsFunc(required []kapi.ResourceName, object runtime.Object) error {
+	if _, ok := object.(*imageapi.ImageStreamTag); !ok {
+		return fmt.Errorf("Unexpected input object %v", object)
```
Comment (smarterclayton) [inline pkg/quota/image/imagestreamtag_evaluator.go]: Errors don't capitalize their messages.

File: pkg/quota/image/imagestreamtag_evaluator.go [hunk 20]
```diff
@@ -0,0 +1,135 @@
...
+				ObjectMeta: kapi.ObjectMeta{
+					Namespace: namespace,
+					Name:      id,
+				},
+			}
+		}
+		return obj, nil
+	}
+
+	return quotautil.NewSharedContextEvaluator(
```
Comment (smarterclayton) [inline pkg/quota/image/imagestreamtag_evaluator.go]: This method has too many arguments, @derekwaynecarr should be a struct

File: pkg/quota/image/imagestreamtag_evaluator.go [hunk 21]
```diff
@@ -0,0 +1,135 @@
...
+// NewImageStreamTagEvaluator computes resource usage of ImageStreamsTags. It's sole purpose is to handle
+// UPDATE admission operations on imageStreamTags resource.
+func NewImageStreamTagEvaluator(osClient osclient.Interface) kquota.Evaluator {
+	computeResources := []kapi.ResourceName{
+		imageapi.ResourceImages,
+	}
+
+	matchesScopeFunc := func(kapi.ResourceQuotaScope, runtime.Object) bool { return true }
+	getFuncByNamespace := func(namespace, id string) (runtime.Object, error) {
+		nameParts := strings.SplitN(id, ":", 2)
```
Comment (smarterclayton) [inline pkg/quota/image/imagestreamtag_evaluator.go]: We have a function to do this, replace.

**Comment by smarterclayton** [comment]
I responded to one of the earlier commits regarding enforcement of image stream tag count.

File: pkg/quota/image/usage.go (hunk 17 above)
Comment (smarterclayton) [inline pkg/quota/image/usage.go]: This is an even bigger problem - you're effectively maintaining a completely custom cache of image streams and images... but this is going to be run in a context where we will already have caches of image streams and images.  So this effectively doubles the memory usage of the controller, which is a significant burden.

This needs to be refactored to share a real client.Cache with at least one other controller in this loop.

File: pkg/quota/image/usage.go [hunk 22]
```diff
@@ -10,139 +13,400 @@ import (
...
-	for _, history := range is.Status.Tags {
+// processImageStreamImages is a utility method that calls a given handler for every image of the given image
+// stream that belongs to internal registry. It process image stream status and optionally spec.
+func (c *GenericImageStreamUsageComputer) processImageStreamImages(
+	is *imageapi.ImageStream,
+	handler func(tag string, dockerImageReference string, image *imageapi.Image) error,
+) error {
+	processedImages := sets.NewString()
+
+	for tag, history := range is.Status.Tags {
```
Comment (smarterclayton) [inline pkg/quota/image/usage.go]: This loop is poorly structured - it would be better to identify a full set of tags (unique) first, group them by their registry, and process them in order.  As is you will fetch the same image multiple times if it is tagged, potentially list the same image streams multiple times, etc.

File: pkg/quota/image/usage.go [hunk 23]
```diff
@@ -10,139 +13,400 @@ import (
...
 			}
 
-			if _, exists := processedImages[imageName]; exists {
-				continue
+			var (
+				err error
+				img *imageapi.Image
+			)
+			if c.fetchImagesFromImageStream {
+				img, err = c.getImageStreamImage(is.Namespace, is.Name+"@"+imageName)
```
Comment (smarterclayton) [inline pkg/quota/image/usage.go]: We should either have a function in helper to do this, or a new one should be added, but no code outside of image/api should ever perform string concatenation to form an ISI or IST name (or parse it, for that matter).

File: pkg/quota/image/usage.go [hunk 24]
```diff
@@ -10,139 +13,400 @@ import (
...
+				if err != nil {
+					glog.Errorf("Failed to get image %s of image stream %s/%s: %v", imageName, is.Namespace, is.Name, err)
```
Comment (smarterclayton) [inline pkg/quota/image/usage.go]: Don't use glog.Errorf() in our code base - if you need to eat an error, use util.HandleError.   However, this is not the right check - you should be specifically ignoring NotFound errors before you log output.

File: pkg/quota/image/usage.go [hunk 25]
```diff
@@ -10,139 +13,400 @@ import (
...
+					continue
+				}
+			} else {
+				img, err = c.getImage(imageName)
+				if err != nil {
+					glog.Errorf("Failed to get image %s: %v", imageName, err)
```
Comment (smarterclayton) [inline pkg/quota/image/usage.go]: Same as above.

File: pkg/quota/image/usage.go [hunk 26]
```diff
@@ -10,139 +13,400 @@ import (
...
+					}
 
-			for _, layer := range img.DockerImageLayers {
-				if _, exists := processedLayers[layer.Name]; exists {
+				} else {
+					tag = ref.Tag
+					if len(tag) == 0 {
+						tag = "latest"
+					}
+					ist, err2 := c.osClient.ImageStreamTags(ref.Namespace).Get(ref.Name, tag)
```
Comment (smarterclayton) [inline pkg/quota/image/usage.go]: This is incredibly inefficient, and we should never be doing this in this code path.

File: pkg/quota/image/usage.go [hunk 27]
```diff
@@ -10,139 +13,400 @@ import (
...
+					if err != nil {
+						glog.Errorf("Failed to get image stream image %s/%s@%s: %v", ref.Namespace, ref.Name, ref.ID, err)
+						continue
+					}
 
-			for _, layer := range img.DockerImageLayers {
-				if _, exists := processedLayers[layer.Name]; exists {
+				} else {
+					tag = ref.Tag
+					if len(tag) == 0 {
```
Comment (smarterclayton) [inline pkg/quota/image/usage.go]: We have functions that do this.

File: pkg/quota/image/usage.go [hunk 28]
```diff
@@ -10,139 +13,400 @@ import (
...
+		}
+
+		ns := namespace
+		if len(objRef.Namespace) > 0 {
+			ns = objRef.Namespace
+		}
+
+		isName := nameParts[0]
+		is, err := c.getImageStream(ns, isName)
+		if err != nil {
```
Comment (smarterclayton) [inline pkg/quota/image/usage.go]: Should be ignoring NotFound

File: pkg/quota/registry.go [hunk 29]
```diff
@@ -8,6 +8,13 @@ import (
...
 // NewRegistry returns a registry object that knows how to evaluate quota usage of OpenShift resources.
-func NewRegistry(osClient osclient.Interface) kquota.Registry {
-	return image.NewImageRegistry(osClient)
+// Registry for resource quota controller cannot be used with resource quota admission plugin and
+// vice versa. If the registry will be used in admission plugin, pass true to forAdmission.
+// See a package documentation of pkg/quota/image for more details.
+func NewRegistry(osClient osclient.Interface, forAdmission bool) kquota.Registry {
+	if forAdmission {
+		return image.NewImageRegistryForAdmission(osClient)
+	} else {
```
Comment (smarterclayton) [inline pkg/quota/registry.go]: No else blocks when you return in the if.

**Comment by smarterclayton** [comment]
On a large set of images and image streams, what performance impact does enabling this have?  Given the number of outstanding issues, the fact that we aren't sharing a cache, and the scope of the change, I'm not sure this is ready for use.  I'd like an assessment of the memory and request impact when we have >100 images or image streams (this looks O(n^2) in places) and a follow up on the lines of the refactoring I mentioned, specifically the reuse of caches.

**Comment by smarterclayton** [comment]
Judging by the queries you have to make in order for this to work, I'm starting to consider whether we need to implement the indices that stef requested for layer and image lookup per namespace and have the quota evaluator call that for calculating # of unique images per namespace.

**Comment by smarterclayton** [comment]
Also, see my comment to the older commit that given the problems with this check in practice (that you don't know for sure on virtual resources whether they will result in a count change) we should not try to be precise here - have quota replenish consult for an accurate check *post* creation, and then simply fail if you are over quota (meaning you could get to some number of images more than that, but in practice it's almost impossible).

**Comment by smarterclayton** [comment]
If no quota references this value, will we still construct these caches?  If so, I think 1.2 needs to disable this by default and add a flag somewhere (admission?) to allow opt in enablement.

File: pkg/quota/image/usage.go (hunk 14 above)
Comment (miminar) [inline pkg/quota/image/usage.go]: Will do.

File: pkg/quota/image/usage.go (hunk 16 above)
Comment (miminar) [inline pkg/quota/image/usage.go]: The one I found wasn't exported and I lacked enough courage to do so. I'll do better this time :).

File: pkg/quota/image/usage.go (hunk 17 above)
Comment (miminar) [inline pkg/quota/image/usage.go]: This cache is always thrown away after use. It's recreated on each admission check and resource quota sync. The admission needs to work with the most up to date object available and don't attempt to update it until the check is finished. I don't think this use case allows for sharing cache with the controller.

File: pkg/quota/image/usage.go (hunk 18 above)
Comment (miminar) [inline pkg/quota/image/usage.go]: Shall we start storing registry urls in etcd? This is not the first place I needed to somehow fetch registry url. It always ends up with parsing some `DockerImageReference`. The problem with these references is that they may get outdated.

In the follow up, I will move the cache to the image registry (which will make it effectively duplicated). But having it in one place easily accessible from anywhere would be really nice.

File: pkg/quota/image/imagestreamtag_evaluator.go (hunk 19 above)
Comment (miminar) [inline pkg/quota/image/imagestreamtag_evaluator.go]: Oops.

File: pkg/quota/image/imagestreamtag_evaluator.go (hunk 20 above)
Comment (miminar) [inline pkg/quota/image/imagestreamtag_evaluator.go]: This was actually written by me. I'll remove the factory.

File: pkg/quota/image/imagestreamtag_evaluator.go (hunk 21 above)
Comment (miminar) [inline pkg/quota/image/imagestreamtag_evaluator.go]: Sure.

File: pkg/quota/image/usage.go (hunk 22 above)
Comment (miminar) [inline pkg/quota/image/usage.go]: That's why the cache is utilized. But your idea makes sense and will improve readibility.

File: pkg/quota/image/usage.go (hunk 23 above)
Comment (miminar) [inline pkg/quota/image/usage.go]: I'll export the helpers and make use of them in the follow-up.

File: pkg/quota/image/usage.go (hunk 26 above)
Comment (miminar) [inline pkg/quota/image/usage.go]: I'll fetch the IS (possibly cached) here and process it with a helper.

**Comment by miminar** [comment]
@smarterclayton
> On a large set of images and image streams, what performance impact does enabling this have?

As mentioned earlier, the caches are thrownaway right after the use. The cache lasts only until an admission check or resource quota controller sync is finished.

Having 10 image streams in the namespace, each having 8 images plus 2 pointing to other project (each time different) will result in:

- 1 list ImageStreams request
- (8*10 + 2*10) get Image requests
- 2*10 get ImageStream requests (in reality many of istags or isimages will refer to some common IS, so the O(n) << n).

When a single IS is being updated either with a CREATE ImageStreamMapping, UPDATE ImageStreamTag, or CREATE/UPDATE ImageStream.

Which is a lot indeed. It would make sense to share image cache because Image objects don't change that much and it would safe us a lot of bandwidth.

> I'm starting to consider whether we need to implement the indices that stef requested for layer and image lookup per namespace and have the quota evaluator call that for calculating # of unique images per namespace.

:+1:

> have quota replenish consult for an accurate check post creation, and then simply fail if you are over quota

I think the precise admission check is fine of ImageStreamImage references. ImageStreamTag should be IMHO treated the way you suggest.

> If no quota references this value, will we still construct these caches?

No, caches are one shot only and will be constructed for some project only if the `openshift.io/images` quota is defined and IS is being created or modified.

File: pkg/quota/image/usage.go (hunk 18 above)
Comment (smarterclayton) [inline pkg/quota/image/usage.go]: The rough plan here is that all "internal" images will be represented by a
stub value, and it should be clear from the API value that the field is
pointing to an image inside of OpenShift (or give the client the tools to
handle that).
